net: Add -networkactive option

hebasto opened the pull request:
Some Bitcoin Core activity is completely local (offline), e.g., reindexing.

The `setnetworkactive` RPC command is already present. This PR adds the corresponding command-line argument / config option, and allows to start the client with disabled p2p network by providing `-networkactive=0` or `-nonetworkactive`.

This was done while reviewing #16981.

hebasto committed:
net: Add -networkactive option

The `setnetworkactive' RPC command is already present.
This new option allows to start the client with disabled p2p network
activity for testing or reindexing.

hebasto committed:
net: Log network activity status change unconditionally

MarcoFalke commented:
There was an idea to overload `-noconnect` with this. https://github.com/bitcoin/bitcoin/issues/15900#issuecomment-487139568

No opinion on that, though.

Concept ACK on adding the functionality.

LarryRuane commented on src/init.cpp:
```diff
@@ -456,6 +456,7 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-proxy=<ip:port>", "Connect through SOCKS5 proxy, set -noproxy to disable (default: disabled)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-proxyrandomize", strprintf("Randomize credentials for every proxy connection. This enables Tor stream isolation (default: %u)", DEFAULT_PROXYRANDOMIZE), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-seednode=<ip>", "Connect to a node to retrieve peer addresses, and disconnect. This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-networkactive", "Enable all p2p network activity (default: 1). Could be changed by the `setnetworkactive' RPC command", ArgsManager::ALLOW_BOOL | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```
The open-quote character before `setnetworkactive` is a back-quote, but the close-quote character is a regular single-quote. There is one other place in this file that uses back-quote characters (addnode), but there they are balanced (back-quote for both open-quote and close-quote).

hebasto replied:
https://www.gnu.org/prep/standards/html_node/Quote-Characters.html

TBH, I'm opposite to using back-quote characters for both opening and closing quotes. If you believe that quotes should be balanced, could they be `'`?

jonatack replied:
ISTM the most common practice ATM is neither quotes nor backticks in the bitcoind help, e.g. ` bitcoind -help | grep "pruneblockchain\|getrawtransaction"`

jonatack replied:
Also: s/Could/Can/

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19473#issuecomment-660453564).

LarryRuane reviewed (commented):
Nice PR! I tested various combinations:
- `bitcoind -networkactive`
- `bitcoind -networkactive=1`
- `bitcoind -networkactive=0`
- `bitcoind -nonetworkactive`

and the behavior was as expected (the first two enabled and the second two disabled networking). I also verified that after disabling networking on the command line, I could change it in both directions using the `setnetworkactive` RPC.

I think it's okay that there's no unit test, because this is test code anyway, and there's no easy place to add one. The only thing bad that could happen is networking isn't enabled by default, and that would be noticed right away!

If you can fix that tiny quote problem, I'd be happy to approve.

hebasto commented:
@LarryRuane 
> The only thing bad that could happen is networking isn't enabled by default, and that would be noticed right away!

I think it _is_ enabled by default :)

LarryRuane commented:
> I think it _is_ enabled by default :)

Yes, sorry, my comment wasn't clear; I just meant that without a test, a bug in this PR could conceivably make networking _disabled_ by default. But, as I said, that would be pretty obvious, so I'm not worried! Plus, now that I think about it, there are numerous existing tests that would fail. So never mind. I was just trying to support your decision not to include a test.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1372,7 +1373,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node)
     node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", &uiInterface, gArgs.GetArg("-bantime", DEFAULT_MISBEHAVING_BANTIME));
     assert(!node.connman);
-    node.connman = std::unique_ptr<CConnman>(new CConnman(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max())));
+    node.connman = std::unique_ptr<CConnman>(new CConnman(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), gArgs.GetBoolArg("-networkactive", true)));
```
unrelated style-nit: Can't this use MakeUnique?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19473#issuecomment-660453564).

MarcoFalke commented on src/init.cpp:
```diff
@@ -456,6 +456,7 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-proxy=<ip:port>", "Connect through SOCKS5 proxy, set -noproxy to disable (default: disabled)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-proxyrandomize", strprintf("Randomize credentials for every proxy connection. This enables Tor stream isolation (default: %u)", DEFAULT_PROXYRANDOMIZE), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-seednode=<ip>", "Connect to a node to retrieve peer addresses, and disconnect. This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-networkactive", "Enable all p2p network activity (default: 1). Could be changed by the `setnetworkactive' RPC command", ArgsManager::ALLOW_BOOL | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```
why is it debug_only?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19473#issuecomment-660453564).

maflcko reviewed (approved):
Tested in the gui:

```
$ ./src/qt/bitcoin-qt -? |grep -A1 networka
  -networkactive
       Enable all p2p network activity (default: 1). Could be changed by the
       setnetworkactive RPC command
```

![Screenshot from 2020-07-18 08-25-05](https://user-images.githubusercontent.com/6399679/87846469-64e98600-c8d0-11ea-9968-80b92b106432.png)


ACK e9a9c9712f 🎈

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e9a9c9712f 🎈
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiw8wwAhcByYIf/MRTpljoUpnyyQOh2M2qMLBv5ECseJmHeC9Tq1n/04FTVnguB
C6AjlPujp1jmIpyA1QLMp5j7hMjjBeRSE9t/xH5Tt1Lc5NpzmYucK6/JXU8+P7jN
EB/FpW2QAsAIBc7Hqyh+Cjq4aJegeNQn/j0zBgGKoiEjZtw2wSNgeOZ4QfezXGFT
iUJrlszz+FYQY4lPijAjIJcyVtWwMNuNG4N1T/gC9fi7vF8sZZCtyDdgZgl4VmDe
0IRKXKRz3HCcSe4o8lkB4AnEzqhhUwyORziMcxqkj8eNDYNhzDCfnHiRXZAnQYbk
whHxEHq68P6eRbnbPy+9MSs3HsE9Gu9jLZavZ6sjzdhHkqQXz06J6Irku/2hvEPa
oGygTwLyv2cvpP1kOW8FoE6ddpCCrCFWC9TkQZyJRhnTsoShcxAix5cp38VEVz9c
0KZNkNQZNZC1xsjN8RKkYKbXi72OziL9dPr2oX1ISTdbrzz+c2u7cP6cuo5a5oty
HclCb1NB
=2rnx
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `289aaa99e8660b4dd3573b6eabcfd374fdeaf85e793ebd854168b6684124745b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108289aaa99e8660b4dd3573b6eabcfd374fdeaf85e793ebd854168b6684124745bf01016092da2937a3dae1e2ffb2a227843e308fff01074460d6d1edaca722602241b564c22c308f1045f129637f008be1e47900d0fd7e70083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108d3d89b42837312e1cb3503efcf17d5e08f1045f129637f00865ad140874a6297b0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010e193ae00f42eb4a79557dac88f5db99a08f1045f129637f0083890a1538d7f2eb30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fe983028d23c01ef1614ffcf68a80d9e08f1045f129637f008a528a0a208777fee0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>


Though, I used the following diff:

```diff
diff --git a/src/init.cpp b/src/init.cpp
index a726c5e062..615f3929d4 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -455,7 +455,7 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-proxy=<ip:port>", "Connect through SOCKS5 proxy, set -noproxy to disable (default: disabled)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-proxyrandomize", strprintf("Randomize credentials for every proxy connection. This enables Tor stream isolation (default: %u)", DEFAULT_PROXYRANDOMIZE), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-seednode=<ip>", "Connect to a node to retrieve peer addresses, and disconnect. This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    gArgs.AddArg("-networkactive", "Enable all p2p network activity (default: 1). Could be changed by the `setnetworkactive' RPC command", ArgsManager::ALLOW_BOOL | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-networkactive", "Enable all p2p network activity (default: 1). Could be changed by the setnetworkactive RPC command", ArgsManager::ALLOW_BOOL, OptionsCategory::CONNECTION);
     gArgs.AddArg("-timeout=<n>", strprintf("Specify connection timeout in milliseconds (minimum: 1, default: %d)", DEFAULT_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-peertimeout=<n>", strprintf("Specify p2p connection timeout in seconds. This option determines the amount of time a peer may be inactive before the connection to it is dropped. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-torcontrol=<ip>:<port>", strprintf("Tor control port to use if onion listening enabled (default: %s)", DEFAULT_TOR_CONTROL), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
@@ -1373,7 +1373,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node)
     assert(!node.banman);
     node.banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", &uiInterface, gArgs.GetArg("-bantime", DEFAULT_MISBEHAVING_BANTIME));
     assert(!node.connman);
-    node.connman = std::unique_ptr<CConnman>(new CConnman(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), gArgs.GetBoolArg("-networkactive", true)));
+    node.connman = MakeUnique<CConnman>(GetRand(std::numeric_limits<uint64_t>::max()), GetRand(std::numeric_limits<uint64_t>::max()), gArgs.GetBoolArg("-networkactive", true));
     // Make mempool generally available in the node context. For example the connection manager, wallet, or RPC threads,
     // which are all started after this, may use it from the node context.
     assert(!node.mempool);

jonatack commented:
Mind adding test coverage?

hebasto committed:
test: Add test coverage for -networkactive option

hebasto commented:
Updated e9a9c9712f8a9f22d90057f095676a29a2d0b77f -> a3232493378d87821a3e3b1c11f9070329cb6c5e ([pr19473.01](https://github.com/hebasto/bitcoin/commits/pr19473.01) -> [pr19473.02](https://github.com/hebasto/bitcoin/commits/pr19473.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19473.01..pr19473.02)):

- addressed all comments

laanwj commented on src/init.cpp:
```diff
@@ -456,6 +456,7 @@ void SetupServerArgs(NodeContext& node)
     gArgs.AddArg("-proxy=<ip:port>", "Connect through SOCKS5 proxy, set -noproxy to disable (default: disabled)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-proxyrandomize", strprintf("Randomize credentials for every proxy connection. This enables Tor stream isolation (default: %u)", DEFAULT_PROXYRANDOMIZE), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     gArgs.AddArg("-seednode=<ip>", "Connect to a node to retrieve peer addresses, and disconnect. This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    gArgs.AddArg("-networkactive", "Enable all p2p network activity (default: 1). Can be changed by the setnetworkactive RPC command", ArgsManager::ALLOW_BOOL, OptionsCategory::CONNECTION);
```
Please capitalize P2P

hebasto replied:
[Indeed](https://github.com/bitcoinops/bitcoinops.github.io/blob/master/STYLE.md#unintroduced-abbreviations)!

[Updated](https://github.com/bitcoin/bitcoin/pull/19473#issuecomment-662666625).

laanwj commented on src/net.cpp:
```diff
@@ -2262,8 +2262,10 @@ void CConnman::SetNetworkActive(bool active)
+CConnman::CConnman(uint64_t nSeed0In, uint64_t nSeed1In, bool network_active)
+    : nSeed0(nSeed0In), nSeed1(nSeed1In)
 {
+    SetNetworkActive(network_active);
```
I don't think it makes any difference in practice in this case, but isn't the general convention to call functions like this on the object only after `Init`?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19473#issuecomment-662666625).

laanwj commented:
utACK otherwise

hebasto commented:
Updated a3232493378d87821a3e3b1c11f9070329cb6c5e -> 2aac093a3d60e446b85eebdf170ea6bed77bec92 ([pr19473.02](https://github.com/hebasto/bitcoin/commits/pr19473.02) -> [pr19473.03](https://github.com/hebasto/bitcoin/commits/pr19473.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19473.02..pr19473.03)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19473#discussion_r459026884):
> Please capitalize P2P

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19473#discussion_r459029798):
> I don't think it makes any difference in practice in this case, but isn't the general convention to call functions like this on the object only after `Init`?

LarryRuane reviewed (approved):
ACK 2aac093a3d60e446b85eebdf170ea6bed77bec92

MarcoFalke commented:
re-ACK 2aac093a3d60e446b85eebdf170ea6bed77bec92 🏠

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 2aac093a3d60e446b85eebdf170ea6bed77bec92 🏠
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgzggv8Cu5uiAMd9wd9hW4BXGop+YSDO/jU8KwnxV9CWvYJ9DRLfI8ke5DtsY9x
yJFXnsihcsHmfXuIwfUrXTwkDQJkMY6iULwXCNPApEBjVPgbzpp1oDUU1YXVgQTW
cZehgGiH5K3O/AGr4PgGQ5+RssYw4l9Q5OO9jB1p9kjnAsKOuwGh5ArWLsLeM/84
vkBXs21DuyzqQpJBGHrqRqNM2kMaoBPmkb7gI5C5sW30ZuoAwN4x2NeqMc0UCsGt
gx3MnOHt1lwe4I75NGHO6VajdKJgYxGiRDCKgBJSFUaaRqrai6u27PBqiuJOBm1z
Cw4emfaTovnLM09LHuBYvDnpXd9wLySRpDe254aKMqett5uI4mKwobs521v40zeA
ETHn/8m6aARqgQKK3H6a61Ts8FBfL3rVFIakAIgC/ncUTEEWa9lhkj4OUTU1Y2fO
k2A9l0TQPm41tFegIGdG0ff9NHDj5urBD/1DXWpVCV0FBxJFxfCof+GfcykYoQLa
RD3CPEnA
=Vohr
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `79e733808b8f76828a454a01b6ad120aa8619b7cef190ba12821d040e4881d01  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010879e733808b8f76828a454a01b6ad120aa8619b7cef190ba12821d040e4881d01f0108c0a577ba17b5c72ef3cd857ef157a6408fff0101e5074d652ac0c1cd653d820814fb5c508f1045f19bb97f0087ebd5e03b6f1a4660083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01052c555d04b39a220baa4f0f0e844069008f1045f19bb96f008b77747f8dd6dee110083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0107de92833c76f9c3788358fd839a1274a08f1045f19bb96f008b35a8c8fcbc6112d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010a9825458a12e30f1a387654dc2a39de008f1045f19bb97f0087f1f6bd18854aafe0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
